core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1244,6 +1245,74 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     */
   def mask: Stream[F, O] = this.handleErrorWith(_ => Stream.empty)
 
+  /**
+    * Like [[Stream.flatMap]] but interrupts the inner stream when new elements arrive in the outer stream.
+    *
+    * Finializers of each inner stream are guaranteed to run before the next inner stream starts.
+    *
+    * When the outer stream stops gracefully, the currently running inner stream will continue to run.
+    *
+    * When an inner stream terminates/interrupts, nothing happens until the next element arrives
+    * in the outer stream(i.e the outer stream holds the stream open during this time or else the
+    * stream terminates)
+    *
+    * When either the inner or outer stream fails, the entire streams fails.
+    *
+    */
```

Comment:
Ok the logic is really starting to appear now :)
That makes it a lot easier to review, and I hope you don't mind all the nitpicking, but this is fairly tricky so I'd like to get it in the best shape possible.